File: 16-trees2/5-balancing-and-rotations/main.cpp
// Tree Rotations - C++ Example

#include <iostream>

struct TreeNode {
    int key;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int k) : key(k), left(nullptr), right(nullptr) {}
};

TreeNode* leftRotate(TreeNode* x) {
    TreeNode* y = x->right;
    x->right = y->left;
    y->left = x;
    return y;
}

TreeNode* rightRotate(TreeNode* y) {
    TreeNode* x = y->left;
    y->left = x->right;
    x->right = y;
    return x;
}

int main() {
    TreeNode* root = new TreeNode(10);
    root->right = new TreeNode(20);
    root = leftRotate(root);
    std::cout << "Root after left rotation: " << root->key << std::endl;
    return 0;
}

File: 16-trees2/5-balancing-and-rotations/main.py
# Tree Rotations - Python Example

class TreeNode:
    def __init__(self, key):
        self.key = key
        self.left = None
        self.right = None

def left_rotate(x):
    y = x.right
    x.right = y.left
    y.left = x
    return y

def right_rotate(y):
    x = y.left
    y.left = x.right
    x.right = y
    return x

# Usage
root = TreeNode(10)
root.right = TreeNode(20)
root = left_rotate(root)
print("Root after left rotation:", root.key)
